docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+
+```ts {1,5-7}
+import { path } from '@kitbag/router'
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+  }),
+})
+```
+
+:::info Route Matching
+When using a param type, if the url cannot be parsed as the specified type it will not be matched. For example if the url was `/events/two-thousand-and-twenty-four/september` it would not be matched because `two-thousand-and-twenty-four` is not a number.
+:::
+
+### Built-in Param Types
+Kitbag Router comes with a few built-in param types:
+
+| Type     | Default | Description                                                                                   |
+| -------- | ------- | --------------------------------------------------------------------------------------------- |
+| `String` | ✓       | Any string value                                                                              |
+| `Number` |         | Any value that can be parsed with the `Number` constructor                                    |
+| `Boolean`|         | The literal values `"true"` or `"false"`                                                      |
+| `Date`   |         | Any value that can be parsed with the `Date` constructor. Uses `toISOString` when serializing |
+| `RegExp` |         | A literal regex expression                                                                    |
+| `JSON`   |         | Any value that can be parsed with the `JSON.parse` method                                     |
+
+## Custom Param Types
+You can define your own param types using the `createParam` utility. This serves three purposes:
+1. Define the type of the param.
+2. Define how the param should be serialized and deserialized.
+3. Determine if a value is valid.
+
+Using custom param types you can define more complex validation rules. For example, this month param must be a valid month name. If it is not, the route will not match.
+
+```ts {3-11,17}
+import { createParam, createRoute, path } from '@kitbag/router'
+
+const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'] as const
+
+const monthParam = createParam((value, { invalid }) => {
+  if (months.includes(value as any)) {
+    return value
+  }
+
+  invalid(`Invalid month: ${value}`)
+})
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: monthParam,
+  }),
+})
+```
+
+By default custom params will use `toString` when serializing. But you can use a Get/Set param to define how the param should be serialized and deserialized. For example we can make the month param capitalized when parsing and lowercase when serializing.
+
+```ts
+import { createParam } from '@kitbag/router'
+
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'] as const
```

Comment:
same "issue" here as above, so whatever you decide there should be applied here as well

Reply:
Updated in https://github.com/kitbagjs/router/pull/395/commits/cb79ed88c5cd4659a79d30ada19f49dba618d4c3